### ISSUE TITLE
User login/sign-up error - "The change you wanted was rejected."

### ISSUE THREAD

**Issue by cloopadoop** [astuto/astuto#201]
1. Signed in with default user.
2. Changed user information.
3. Restarted server, found that the service did not restart. Restarted service.
4. Was able to log in with default credentials but not renamed credentials.
5. Attempted to set docker container to restart (still no luck on that front.)
7. Logging in with default credentials or renamed credentials or creating a new user yields this error:
![image](https://user-images.githubusercontent.com/1791012/221031196-8dd62be3-1cb4-4d15-b4c6-e2d71eb036da.png)

**Comment by cloopadoop** [comment]
It would appear my issue was an Apache2 setup that was rewriting all URL requests to use HTTPS, which apparently is not supported by astuto!

Still can't seem to get the containers to start up automatically when the VPS power cycles, though.

Also, this is very cool. :)

**Comment by sam-morin** [comment]
@cloopadoop 
I am also having this issue. I'm behind an nginx reverse proxy in a docker container. I am on HSTS TLD so plain HTTP is not an option. 

Based on your comment (thank you), after adding: 
`proxy_set_header X-Forwarded-Proto $scheme;`
to my server block, everything seems to be functioning as expected now.

I don't use Apache much but try adding:
`RequestHeader set X-Forwarded-Proto $scheme`
to your VirtualHost and see if that resolves the issue.

**Comment by riggraz** [comment]
Hi guys, thanks for reaching out!

Unfortunately, I never run Astuto over SSL either, so right now I can't be much of help. However, I plan on releasing a new version of the demo that runs on https pretty soon (probably at the end of next month), so hopefully I'll be able to dig into the problem.

**Comment by riggraz** [comment]
Hi, just a quick update.

@byjrack has posted an issue with useful knowledge for setting up nginx and TLS, you may want to check it out here: https://github.com/astuto/astuto/issues/203#issuecomment-1456842967

**Comment by byjrack** [comment]
Yeah I feel like I saw that error more than a couple times when I was figuring out a TLS termination model myself.  I don't have an HTTPD config for the reverse, but it shouldn't be too far off of the nginx config.  My gut says it's likely a callback looping through the client that may require an 80/HTTP redirect handler to work correctly.

**Comment by DockDockGoat** [comment]
I'm having this issue as well. I thought I'd post my workaround for others to see.

My setup is that I have the containers running behind an haproxy instance, where I terminate all my HTTPS connections, then haproxy talks to the backend over HTTP. The load balancer also redirects all inbound HTTP traffic to HTTPS. The redirect part in haproxy.cfg is:

`http-request redirect scheme https unless { ssl_fc }`

I build the containers with this docker-compose.yml:

```
services:
  db:
    image: postgres:14.5
    environment: &db-env
      POSTGRES_USER: astuto
      POSTGRES_PASSWORD: verygoodpassword
    volumes:
      - dbdata:/var/lib/postgresql/data
  web:
    image: riggraz/astuto:latest
    environment:
      <<: *db-env
      BASE_URL: https://astuto.example.com
      SECRET_KEY_BASE: verygoodsecretkey
      RAILS_LOG_TO_STDOUT: true
    ports:
      - "3000:3000"
    depends_on:
      - db
    
volumes:
  dbdata:
```

When I try and log in with the admin user, I get the "change you wanted was rejected" error, and I see this in the console logs, suggesting it's an HTTP/HTTPS mismatch issue:

```
web-1  | W, [2025-10-16T21:50:30.553570 #33]  WARN -- : [dfb22182-1abf-40f1-936e-2991925d6188] [http://astuto.example.com/users/sign_in] [10.1.3.24] HTTP Origin header (https://astuto.example.com) didn't match request.base_url (http://astuto.example.com)
web-1  | I, [2025-10-16T21:50:30.554377 #33]  INFO -- : [dfb22182-1abf-40f1-936e-2991925d6188] [http://astuto.example.com/users/sign_in] [10.1.3.24] Completed 422 Unprocessable Entity in 9ms (ActiveRecord: 1.6ms | Allocations: 1544)
web-1  | F, [2025-10-16T21:50:30.555899 #33] FATAL -- : [dfb22182-1abf-40f1-936e-2991925d6188] [http://astuto.example.com/users/sign_in] [10.1.3.24]      
web-1  | [dfb22182-1abf-40f1-936e-2991925d6188] [http://astuto.example.com/users/sign_in] [10.1.3.24] ActionController::InvalidAuthenticityToken (ActionController::InvalidAuthenticityToken):
```

I've rebuilt the containers a bunch of times with different **BASE_URL** environment values, and the entry in the console logs never changes. I can see the environment variable gets passed to the container when I check with:

```
root@verygoodhostname:~# docker exec -it 21cf67deb4f0 env |grep BASE_URL
BASE_URL=https://astuto.example.com
```

I'm not sure how to diagnose astuto/rails/puma further, but the BASE_URL env variable doesn't seem to be getting into the app in a meaningful way. It looks to me like it's completely ignoring that and instead generating it from the HTTP request headers. Changing not only http or https, but the dns hostname completely never shows up in the console logs.

The workaround I did is to fake the origin in the load balancer. In haproxy.cfg:

```
  # tell astuto/rails/puma there's no https
  acl ACL_astuto hdr(host) -i astuto.example.com
  http-request set-header origin http://astuto.example.com if ACL_astuto
```

After setting that header, it works as expected. My take is that there's a bug somewhere in the stack where the BASE_URL environment variable isn't making it into the app. I've never had to set an origin header for any other web apps behind a load balancer like this, but this workaround does appear to work in my environment.
